<a href="https://colab.research.google.com/github/lugsantistebanji/WCS-IA/blob/main/WCS_IA_Quetes_API_JSON_to_DataFrame.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# __Parse a JSON file with Pandas__

In [8]:
import json
import requests
import pandas as pd


link = "https://raw.githubusercontent.com/WildCodeSchool/java-jsonparse/master/weather.json"

# This function gets the file content from the URL
r = requests.get(link)

# You can access to the text content with this function
r.text

data = json.loads( r.text)

In [2]:
data

{'coord': {'lon': -0.13, 'lat': 51.51},
 'weather': [{'id': 300,
   'main': 'Drizzle',
   'description': 'light intensity drizzle',
   'icon': '09d'},
  {'id': 800, 'main': 'Clear', 'description': 'clear sky', 'icon': '01n'}],
 'base': 'stations',
 'main': {'temp': 280.32,
  'pressure': 1012,
  'humidity': 81,
  'temp_min': 279.15,
  'temp_max': 281.15},
 'visibility': 10000,
 'wind': {'speed': 4.1, 'deg': 80},
 'clouds': {'all': 90},
 'dt': 1485789600,
 'sys': {'type': 1,
  'id': 5091,
  'message': 0.0103,
  'country': 'GB',
  'sunrise': 1485762037,
  'sunset': 1485794875},
 'id': 2643743,
 'name': 'London',
 'cod': 200}

---

Voici le résultat attendu de ces 3 premières questions :

* City name: London
* City latitude: 51.51
* City longitude: -0.13
* Weather: Drizzle
* Weather: Clear

___


1. __Affiche la valeur de l'attribut name__

In [4]:
print(f"City name: {data['name']}")

City name: London


___

2. __Affiche les valeurs des attributs lat et lon de l'élément coord contenu dans la racine__


In [5]:
print(f"""City latitude: {data['coord']['lat']}
City longitude: {data['coord']['lon']}
""")

City latitude: 51.51
City longitude: -0.13



___

3. __Parcours tous les éléments dans weather et affiche le contenu de main (peut-être avec une boucle ?)__

In [6]:
for weather in data['weather']:
    print(f"Weather: {weather['main']}")

Weather: Drizzle
Weather: Clear


___

4. __Essaye maintenant d'importer ton fichier json dans pandas :__

    You can directly read JSON from your request object

In [7]:
print(r.json())

{'coord': {'lon': -0.13, 'lat': 51.51}, 'weather': [{'id': 300, 'main': 'Drizzle', 'description': 'light intensity drizzle', 'icon': '09d'}, {'id': 800, 'main': 'Clear', 'description': 'clear sky', 'icon': '01n'}], 'base': 'stations', 'main': {'temp': 280.32, 'pressure': 1012, 'humidity': 81, 'temp_min': 279.15, 'temp_max': 281.15}, 'visibility': 10000, 'wind': {'speed': 4.1, 'deg': 80}, 'clouds': {'all': 90}, 'dt': 1485789600, 'sys': {'type': 1, 'id': 5091, 'message': 0.0103, 'country': 'GB', 'sunrise': 1485762037, 'sunset': 1485794875}, 'id': 2643743, 'name': 'London', 'cod': 200}


And you can import JSON object to pandas

In [9]:
pd.json_normalize(r.json())

weather      base  visibility  \
0  [{'id': 300, 'main': 'Drizzle', 'description':...  stations       10000   

           dt       id    name  cod  coord.lon  coord.lat  main.temp  ...  \
0  1485789600  2643743  London  200      -0.13      51.51     280.32  ...   

   main.temp_max  wind.speed  wind.deg  clouds.all  sys.type  sys.id  \
0         281.15         4.1        80          90         1    5091   

   sys.message  sys.country  sys.sunrise  sys.sunset  
0       0.0103           GB   1485762037  1485794875  

[1 rows x 23 columns]

Tu obtiens un DataFrame avec une seule ligne ? C'est normal, notre fichier JSON contient un premier niveau "weather". Nous allons indiquer à Pandas de descendre au niveau inférieur (ici dans l'enregistrement weather). Pour cela, nous utilisons le paramètre record_path. Nous pouvons aussi utiliser le paramètre meta pour indiquer de conserver des éléments d'un niveau supérieur:

In [10]:
pd.json_normalize(r.json(), record_path= "weather", meta = 'dt')

id     main              description icon          dt
0  300  Drizzle  light intensity drizzle  09d  1485789600
1  800    Clear                clear sky  01n  1485789600

___

5. __Filtre maintenant ton DataFrame avec Pandas__

Afficher uniquement les colonnes suivantes : main (celui qui est à l'intérieur de weather), name, and visibility.

In [18]:
pd.json_normalize(r.json(), record_path= "weather", meta = ['name', 'visibility'])[['name','main','visibility']]

name     main visibility
0  London  Drizzle      10000
1  London    Clear      10000

___